syntax = "proto3";

package lightdfs;

service NameServer
{
    // 注册数据服务器
    rpc RegisterDataServer(DataServerInfo) returns (Response);
    // 获取数据服务器列表
    rpc GetDataServerList(empty) returns (GetDataServerListResponse);
    // 下线数据服务器
    rpc LogoutDataServer(DataServerInfo) returns (Response);
    // 注册用户
    rpc RegisterUser(RegisterRequest) returns (Response);
    // 登录
    rpc Login(LoginRequest) returns (LoginResponse);
    // 文件锁，lock_type: 0:读锁，1:写锁
    rpc LockFile(LockFileRequest) returns (Response);
    rpc UnlockFile(UnlockFileRequest) returns (Response);
    // 检查缓存
    rpc CheckCache(CheckCacheRequest) returns (Response);
    // 添加新文件（夹）,提供完整文件信息
    rpc AddFile(FileInfo) returns (Response);
    // 删除文件（夹），提供文件路径
    rpc DeleteFile(DeleteRequest) returns (Response);
    // 修改文件（夹），提供原始文件路径，新文件路径，新文件大小，修改时间
    rpc ModifyFile(ModifyFileRequest) returns (Response);
    // 获取文件信息， 提供文件路径
    rpc GetFileInfo(GetFileInfoRequest) returns (FileInfoResponse);
    // 验证JWT
    rpc VerifyJWT(VerifyJWTRequest) returns (Response) {}
}

message Response
{
    int32 success  = 1;
    string message = 2;
}

message DataServerInfo
{
    int64 id    = 1;
    string host = 2;
    int32 port  = 3;
}

message empty
{
    int32 e = 1;
}

message GetDataServerListResponse
{
    int32 success                              = 1;
    string message                             = 2;
    repeated DataServerInfo dataServerInfoList = 3;
}

message RegisterRequest
{
    string username = 1;
    string password = 2;
}

message LoginRequest
{
    string username = 1;
    string password = 2;
}

message LoginResponse
{
    int32 success  = 1;
    string message = 2;
    string jwt     = 3;
}

message LockFileRequest
{
    int32 lock_type = 1;
    string filepath = 2;
}

message UnlockFileRequest
{
    int32 lock_type = 1;
    string filepath = 2;
}

message CheckCacheRequest
{
    string absolute_path = 1;
    float mtime          = 2;
}

message GetFileInfoRequest
{
    string absolute_path = 1;
}

message FileInfo
{
    string absolute_path = 1;
    int64 size           = 2;
    bool is_dir          = 3;
    float ctime          = 4;
    float mtime          = 5;
}

message DeleteRequest
{
    string absolute_path = 1;
}

message ModifyFileRequest
{
    string old_absolute_path = 1;
    string new_absolute_path = 2;
    int64 new_size           = 3;
    float mtime              = 4;
}

message FileInfoResponse
{
    int64 size     = 1;
    bool is_dir    = 2;
    float ctime    = 3;
    float mtime    = 4;
    int32 success  = 5;
    string message = 6;
}

message VerifyJWTRequest
{
    string jwt = 1;
}